core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleGetRecoveryCandidatesHandler.java

```diff
@@ -0,0 +1,164 @@
+                    startedInstanceMetadataWithServicesNotRunning = collectRecoveryCandidates(startedInstancesMetadata,
+                            detailedHostStatuses);
+                }
+            }
+
+            List<CloudInstance> instancesWithServicesNotRunning;
+
+            if (startedInstanceMetadataWithServicesNotRunning.isEmpty()) {
+                instancesWithServicesNotRunning = Collections.emptyList();
+            } else {
+                instancesWithServicesNotRunning = instanceMetaDataToCloudInstanceConverter.convert(startedInstanceMetadataWithServicesNotRunning,
+                        stack.getStack());
+                LOGGER.info("Collected {} downscale recovery candidates for hostGroup: {}", instancesWithServicesNotRunning.size(), request.getHostGroupName());
+            }
+
+            StopStartDownscaleGetRecoveryCandidatesResult result = new StopStartDownscaleGetRecoveryCandidatesResult(stack.getId(),
+                    instancesWithServicesNotRunning, request.getHostGroupName(), request.getHostIds());
+            notify(result, event);
+        } catch (Exception e) {
+            String errorMessage = String.format("Failed to collect downscale recovery candidates for stack: %s. " +
+                    "Proceeding with subsequent flow.", request.getResourceId());
+            LOGGER.warn(errorMessage, e);
+            StopStartDownscaleGetRecoveryCandidatesResult result = new StopStartDownscaleGetRecoveryCandidatesResult(request.getResourceId(),
+                    Collections.emptyList(), request.getHostGroupName(), request.getHostIds());
+            notify(result, event);
+        }
+    }
+
+    private List<CloudInstance> collectStartedInstancesFromCloudProvider(CloudConnector connector, AuthenticatedContext ac,
+            List<CloudInstance> cloudInstances) {
+        List<CloudVmInstanceStatus> vmInstanceStatuses = connector.instances().checkWithoutRetry(ac, cloudInstances);
+        return vmInstanceStatuses.stream()
+                .filter(vm -> InstanceStatus.STARTED.equals(vm.getStatus()))
+                .map(CloudVmInstanceStatus::getCloudInstance)
+                .collect(Collectors.toList());
+    }
+
+    private List<InstanceMetadataView> collectRecoveryCandidates(List<InstanceMetadataView> startedInstances, DetailedHostStatuses detailedHostStatuses) {
+        Predicate<InstanceMetadataView> instanceHealthCheck =  im -> {
+            HostName hostName = hostName(im.getDiscoveryFQDN());
+            boolean hostUnhealthy = detailedHostStatuses.isHostUnHealthy(hostName);
+            boolean maintenanceMode = detailedHostStatuses.isHostInMaintenanceMode(hostName);
+            boolean hostDecommissioned = detailedHostStatuses.isHostDecommissioned(hostName);
+            boolean servicesUnhealthy = detailedHostStatuses.areServicesUnhealthy(hostName);
+                    // CM: healthy, Services: healthy, Host: Unhealthy
```

Comment:
How does this entire statement map to hostUnhealthy? Likewise for the subsequent conditions mentioned and followed by a single condition.

To map to the writeup, can this be changed to: 
return
// hostHealthy, decommissioned:true, maintMode:true, serviceRunning: false
(!hostUnhealthy && hostDecommissioned && maintenanceMode && serviceNotRunning) || 
// hostHealthy, decommissioned:true, maintMode:false, serviceRunning: false
(!hostUnhealthy && hostDecommissioned && !maintenanceMode && servicesRunning)

Will ofcourse need to be adjusted for the behaviour in scenarios where service health is unavailable.

Reply:
Updated to map write-up